src/EnergyPlus/SurfaceGeometry.cc

```diff
@@ -1346,17 +1344,19 @@ namespace SurfaceGeometry {
 
         // Move all shading Surfaces to Front
 
+        state.dataSurfaceGeometry->SurfaceTmpClassInvalid.dimension(TotSurfaces, false);
+        state.dataSurfaceGeometry->SurfaceTmpClassMoved.dimension(TotSurfaces, false);
```

Comment:
Are these used anywhere outside this function? Could these be local here?

Reply:
They are used in `GetSurfaceData`, `GetIntMassSurfaceData`, `GetRectSurfaces`, `GetHTSurfaceData`. The last three are called by `GetSurfaceData`. I don't know what is better, using them as state variables, or pass them as function parameters?

Reply:
I think Moved is only used in GetSurfaceData, so it can be local. And the other one is fine as a state variable.

Reply:
Or would it make sense to add a new surface class enum for `Invalid`?

Reply:
That one could be last in the list and then you could use it in the range test for that assert `if (>None && <Invalid)`, maybe?

Reply:
I don't know. I was gonna explicitly add a `const int NO_SURFACE_CLASS = 15;`.
First `Invalid` is not literally a `SurfaceClass`. And no implicit way to prevent more classes to append after it. 

Here's [a post](https://stackoverflow.com/questions/2102582/how-can-i-count-the-items-in-an-enum) about using an extra (last) item of an ENUM class to represent its length, as the example below shows. `Invalid` may be used this way, but again it's not a 100% safe solution. 
`enum class Example { A, B, C, D, E, ExampleCount };`

Reply:
No, you shouldn't add a const NO_SURFACE_CLASS just for MOVED. MOVED is temporary. In fact, you can probably just make the MOVED array local to the function that uses it. 

As far as using enum class Example { A, B, C, D, Count }; and then using Example::Count as the length of the array ... that works if A is set to 0.

Reply:
My suggestion was not to replace MOVED, but rather to replace INVALID. Storing INVALID briefly during initial setup is one issue. Counting the length of the enum to size the various reporting arrays is another.

Reply:
Issue 1. INVALID. Is this a piece of information that needs to persist throughout the simulation or does it immediately trigger a fatal error and the simulation ends? If the former, I think you can have INVALID as part of the enum. Just set its value to -1 if you don't want to do any reporting on it. 

Issue 2. Enumeration counts. This is really only useful if the enumeration starts at 0 (or 1) and is dense.  [Ed: if an enumeration doesn't start at 0 (or 1) and/or is not dense, it probably shouldn't be an enumeration.] Using Count (I've usually called it Num, but whatever) is a good shorthand, and is automatic if you are starting the enumeration at 0. If you are starting at 1, you can always define Count to have the same value as the last element in your set, e.g., enum Example { A, B, C, D, Count = D };

Reply:
INVALID is set during input processing for surfaces, and it eventually trips a fatal error. It's used in `GetSurfaceData` and a few functions called from there.

Reply:
As long as the simulation exits when there is an INVALID surface before any array that is indexed by enum SurfaceClass is read or written, it is okay to have INVALID as enum SurfaceClass -1.

Reply:
The [latest commit](https://github.com/NREL/EnergyPlus/pull/8347/commits/83d8eba37e731a54e1cb2728afd00b078f740552) added `SurfaceClassCount` (It's first item is None = 0). Looks like it's reasonable to add INVALID = -1 at the top.

Reply:
You can probably just call it Count, because SurfaceClass is already the name of the enum.